### DISCUSSION TITLE
WebP on-the-fly conversion

### DISCUSSION THREAD

**Discussion by fillito** [anansi-project/opds-pse#10] [General]
A user recently contacted Panels support to see if we could do something to speed up their streaming by tweaking how Panels pre-cache pages. 
Apparently, their whole library of comics is in WebP format, and it takes some time to get the pages converted to jpeg. 
I still need to find out what server they're using.

Panels already has support for WebP, and it's a waste of resources to convert each page on-the-fly when a client has support for the original format. It would also speed the streaming up. 

Does it make sense to have an opt-in feature to allow servers to return the WebP image and skip the conversion? 

I don't have a specific solution in mind. I'm just testing the water. 
What do you think would be a good approach?

**Comment by gotson** [comment]
It's been raised on Komga: https://github.com/gotson/komga/issues/1098

**Comment by gotson** [comment]
OPDS-PSE mandates the use of jpeg, gif or png images, see [here](https://anansi-project.github.io/docs/opds-pse/specs/v1.2#specifications):

> type MUST indicate the type of image returned by the server (value MUST be one of the following : image/jpeg, image/gif or image/png)

I agree with you that this is outdated, and we need a better way to handle that.

> Does it make sense to have an opt-in feature to allow servers to return the WebP image and skip the conversion?

I don't think the setting should be at the server level, but needs to be baked in the OPDS-PSE protocol, and the client should be allowed to request `webp` content.

---

In Komga, when using the REST API, clients would typically:
- get the list of all pages, which would also provide what media type the page is in
- know what formats they can handle
- when requesting a page, the client would request a conversion if the format is not something they can handle

We can't really do that with OPDS-PSE, because the OPDS-PSE link only provides a single mediatype for the whole publication, while each page could be using a different mediatype.

---

One way of doing it would be to use [Content Negociation](https://developer.mozilla.org/en-US/docs/Web/HTTP/Content_negotiation).

**Reply by gotson** [reply]
If we were to use Content Negociation, the OPDS-PSE client could request images by adding the `Accept` header.

That's what browsers do when requesting images, for example Chrome sends:

```
accept: image/avif,image/webp,image/apng,image/svg+xml,image/*,*/*;q=0.8
```

It seems `image/*` is some kind of catch-all used for well-known image types, like jpeg/png/gif.

The OPDS-PSE server behaviour and response could be:
- if the client requested formats contain the native format of the publication, provide it
- if the client requested formats does not contain the native format of the publication, then convert on the fly to one of the requested formats, by following the [quality factor](https://developer.mozilla.org/en-US/docs/Glossary/Quality_values).
- if the client requested formats does not contain the native format of the publication, and the server cannot convert to any format requested, the server returns an error code 406 NOT ACCEPTABLE.

An `Accept` header that would be compatible with the current formats (jpeg, png, gif) should be:

```
accept: image/jpeg, image/png, image/gif ,image/*;q=0.8
```

Which would mean:
- send me jpg, png or gif
- if you can't, send me any kind of image you have

If the client supports `webp`, then it should send:

```
accept: image/webp, image/jpeg, image/png, image/gif ,image/*;q=0.8
```

If we were to implement this, that would render the `type` attribute on the OPDS-PSE link irrelevant. We could create a backwardc-compatible version 1.3 of the specification:
- which would require clients to ignore the `type` attribute
- which would require clients to use Content Negociation instead

That way we don't have to make a breaking-change to the specification.

WDYT @fillito ? Would that work in Panels ?

cc @ajslater @majora2007 @Difegue

**Reply by fillito** [reply]
That would definitely work for us 👍  I'm happy to adopt that solution and start adding the `Accept` header in the requests if it's added to the specs 👌

**Reply by majora2007** [reply]
I am fine with this proposal as well, however do error codes being sent back actually work on clients? I've seen many OPDS clients that cannot handle bad requests or even an OPDS xml with an error message in it. Will this work for anything other than Panels? Will the users understand what to report on?

**Reply by Difegue** [reply]
This sounds good to me as well - I have some converting infrastructure in place already as part of an existing reduce-bandwidth option, so this'd be an easy solution to adopt. 

I believe OPDS clients that don't handle error 406s shouldn't be 1.3 compatible -- But there's no real way to indicate that to users, right? 🤔

**Reply by gotson** [reply]
Indeed, we may need to check what are the accept headers sent by existing clients.

The server won't be able to tell whether the request was made from an 1.2 or below client, or a 1.3 client. 

I won't be available for the next 3 weeks or so, if someone wants to try to see what popular clients are sending in the Accept header, that could be helpful.

I think any opds client getting a 4xx or 5xx would display some kind of "oh snap, the image couldn't be loaded". It could happen for various reasons like bad network or temporary issue.

**Reply by gotson** [reply]
Thinking more about it, the above proposal is transparent for both client and server, which indeed could have side effects.

Thinking out loud, here are some ways to make it more explicit:
1. Add a parameter in the `href` link. There's already some parameters, so we could have an extra one, that would be optional, and if present the server would honor the `Accept` header.
2. Make a version 2 of the opds pse protocol, which would use a different xml namespace. Both could co-exist in the same opds catalog. There would be an additional `link` with a different `rel` and `href`. I don't know how existing opds-pse clients parse the xml though, it could lead to some issues.

**Reply by fillito** [reply]
The more explicit, the better. IMO. 
1. This sounds good to me. 
2. This approach also sounds great, and it involves having less optionality. The only issue I see with this approach is that it "forks" the specs. I assume the plan would be to "deprecate" 1.x versions and continue supporting and evolving the new version. Am I right?   
(Tangent to this topic, but kind of related if there are plans to increase the OPDS-PSE version to the next major; I opened a [new thread about OPDS 2](https://github.com/anansi-project/opds-pse/discussions/11))

**Reply by majora2007** [reply]
I don't think we should do anything that forks the spec. We don't have enough consensus from other client apps that implement the spec. Looking even at that OPDS 2 spec, they have more image support, but don't have only WebP support even in their image spec. They require gif, png, or jpeg always.

**Reply by majora2007** [reply]
I think we need to move this discussion over to https://github.com/opds-community/specs where the official spec lies. The goal isn't to maintain our own spec just for the readers, but to keep everyone in alignment.

**Reply by therobbiedavis** [reply]
They actually host a monthly call that's public invite. Might be beneficial for moving the spec forward for people to attend.
https://docs.google.com/document/d/1759wJIS5pDVQz8fT18hk7tCV2JWMJAnR1sT5uWD9W6I/edit?usp=sharing
https://calendar.google.com/calendar/u/0?cid=NjYwZjRhZDc3YTE0ZDg5MTM2NzExZTBiMjNkNjdhZWYxNzE1NDI2NDkxOGU5ZWI0YWYzMTgzNGY4ZmUyMzI0MUBncm91cC5jYWxlbmRhci5nb29nbGUuY29t

This is probably also a better place for the discussion: https://github.com/opds-community/drafts/discussions

**Reply by HadrienGardeur** [reply]
Content negotiation is an elegant HTTP-based solution to this problem, but it requires an extra round trip:

- the client sends a `HEAD` request to the server
- the server responds with a list of supported formats
- the client then requests the proper format using a `GET` request

If you add other potential hints and properties such as height/width/resolution, it doesn't necessarily scale that well.

My advice would be to look for inspiration in the [IIIF Image API specification](https://iiif.io/api/image/3.0/), move to OPDS 2.0 and define an updated URI template and list of hints embedded in `properties`.

This could look something like that:
```json
{
  "rel": "http://vaemendis.net/opds-pse/stream",
  "href": "https://example.com/opds-pse/xyz{?page,height,width,format}",
  "templated": true,
  "properties": {
    "maxHeight": 1600,
    "maxWidth": 1200,
    "supportedFormats": ["image/webp", "image/jpeg", "image/png", "image/gif"]
  }
}
```

You'll notice that I've skipped the number of pages in those hints. This is potentially redundant with what can be expressed in OPDS 2.0 with the following metadata: https://readium.org/webpub-manifest/contexts/default/#duration-and-number-of-pages

**Reply by gotson** [reply]
> I don't think we should do anything that forks the spec.

We are not talking about forking any spec here. It would create a version 2.0 of the OPDS-PSE, that's not forking anything, given we maintain the OPDS-PSE now.

> Looking even at that OPDS 2 spec, they have more image support, but don't have only WebP support even in their image spec. They require gif, png, or jpeg always.

if i read that correctly, i think it requires at least one link to be gif/png/jpeg, but it should allow for other kind of media type.

> I think we need to move this discussion over to https://github.com/opds-community/specs where the official spec lies. The goal isn't to maintain our own spec just for the readers, but to keep everyone in alignment.

That's probably something we need to discuss with @HadrienGardeur, to understand whether something like page streaming could be located there or not. I am all in favor of moving that unofficial spec to something official if we can.

> Content negotiation is an elegant HTTP-based solution to this problem, but it requires an extra round trip:
> 
> * the client sends a `HEAD` request to the server
> * the server responds with a list of supported formats
> * the client then requests the proper format using a `GET` request

Content negociation can work in 2 different ways:
- server-driven negotiation or proactive negotiation, which is what i propose (no round-trip)
- agent-driven negotiation or reactive negotiation, which is what you describe

**Reply by HadrienGardeur** [reply]
Proactive negotiation can avoid a round trip as long as you accept the risk that the server might not be able to reply with your preferred media type. This risk is mitigated when a long list of media type is specified, which is the case with your example @gotson.

Knowing in advance what the server can return remains useful even in that case, since it allows the client to adapt what they send in their `Accept` header.
Hints also remain useful since you also want to communicate the max height and width available.

With this mixed approach, you could drop the media type from the URI template:

```json
{
  "rel": "http://vaemendis.net/opds-pse/stream",
  "href": "https://example.com/opds-pse/xyz{?page,height,width}",
  "templated": true,
  "properties": {
    "maxHeight": 1600,
    "maxWidth": 1200,
    "supportedFormats": ["image/webp", "image/jpeg", "image/png", "image/gif"]
  }
}
```

> Looking even at that OPDS 2 spec, they have more image support, but don't have only WebP support even in their image spec. They require gif, png, or jpeg always.

The spec requires specific media types for the cover of a publication (and IMO we should extend this requirement to WebP/AVIF) but it doesn't say anything about the content of the publication itself. In this case, using WebP or (even better) AVIF would be perfectly acceptable.

Going beyond a URI template based approach to stream comics/manga, [the Divina profile of RWPM](https://readium.org/webpub-manifest/profiles/divina.html) only requires a bitmap format and doesn't list specific ones.

While Divina remains a work in progress, it has several benefits over a single URI template:

- it includes information about spreads and where each page belongs in a spread
- guided navigation (also known as panel by panel navigation) can be authored in a Divina or pre-processed by the server

It would be very interesting to discuss Divina between both of our communities (Readium and Anansi-Project) since Divina could IMO cover the use cases for OPDS-PSE and ComicInfo.xml at the same time.

**Comment by awh-tokyo** [comment]
> If we were to implement this, that would render the type attribute on the OPDS-PSE link irrelevant.

Unrelated to this issue, but I have books that are made of more than one type of image, and I'd be happy to see the "type" attribute go away because I know that sometimes the server isn't truthful with the client about what to expect.

**Reply by gotson** [reply]
We all agree that the current way of doing this is not optimal, my proposal should tackle this.

**Reply by HadrienGardeur** [reply]
In OPDS 2.0, the `type` attribute becomes optional for links.

If the server allows transcoding on the fly, that's not an issue anyway. 

But if you want to simply stream content "as-is", moving away from a URI template based approach would allow you to provide such info along with other useful metadata.
